HW0/P5/P5.txt

```diff
@@ -1 +1,10 @@
-# Your answers here.
+1) Adding together the numbers takes 255 operations (seconds) since each number has to be added in at some point
+2) It would take 34 seconds. You give each friend 32 numbers to add which takes 31 seconds and produces 8 sums. Then you pair wise add to 4 sums, once again to 2, and finally one more second to get to one. This is a total of 34 seconds. 
+3) If everyone is working it is the log base 2 of 256 which is 8 seconds. This only requires 128 people, any more would be redundant
+4) Like part 3 it's the log base two of N since you add two numbers at a time and combine them and repeat. On the plot the blue is the linear time it takes one person to add the numbers, and the green is the logarithmic time 
+5) This is 15 seconds. It takes twice as long as in part three because after every round of counting, you need exactly one second for half the people to tell the other half. It is 1 less than double because at the end you don't need to tell anyone the final count. 
+6) Assuming it still takes one second to communicate a number, the answer becomes 512. Starting with the simple scenario of two bags, after two seconds you have handed out both bags, third second to communicate a number, fourth second to add. With a third bag, the person has to communicate the number, and an addition has to be made again giving six seconds. The addition of a bag leads to an additional two seconds in every case. 
```

Comment:
The problem with this analysis is that it doesn't scale linearly from the two-bag case. With more bags and more cashiers, different cashiers can be doing different things at the same time, so it doesn't have to take as long as 512 seconds. Each pair of bags might take 4 seconds to process, but those 4 seconds can overlap with the 4 seconds of processing another pair of bags, so the total time in seconds doesn't have to be twice the number of bags.